File: 9.another-language/9.1 student-grade-analyser-c++/student-grade-analyser.cpp
#include "splashkit.h"
#include "utilities.h"

void input_grades(int grades[], int &size)
{
    size = read_integer_range("How many grades (1 - 20): ", 1, 20);

    for(int i = 0; i < size; i++)
    {
        grades[i] = read_integer("Enter grade " + to_string(i + 1) + ": ");
    }
}


double calculate_average(int grades[], int size)
{
    int total = 0;

    for(int i = 0; i < size; i++)
    {
        total += grades[i];
    }

    return (double)total / size;
}


double calculate_pass_rate(int grades[], int size)
{
    int pass_count = 0;

    for(int i = 0; i < size; i++)
    {
        if(grades[i] >= 50)
        {
            pass_count++;
        }
    }

    return ((double)pass_count / size) * 100;
}


int count_hd(int grades[], int size)
{
    int hd_count = 0;

    for(int i = 0; i < size; i++)
    {
        if(grades[i] >= 80)
        {
            hd_count++;
        }
    }

    return hd_count;
}


void display_results(double average, double pass_rate, int hd_count)
{
    write_line("\nResults");
    write_line("Average Grade: " + to_string(average, 2));
    write_line("Pass Rate: " + to_string(pass_rate, 2) + "%");
    write_line("HD Count: " + to_string(hd_count));
}


int main()
{
    int grades[20];
    int size;

    double average;
    double pass_rate;
    int hd_count;

    input_grades(grades, size);

    average = calculate_average(grades, size);
    pass_rate = calculate_pass_rate(grades, size);
    hd_count = count_hd(grades, size);

    display_results(average, pass_rate, hd_count);

    return 0;
}

File: 9.another-language/9.2 student-grade-analyser-python/student-grade-analyser.py
def input_grades():
    grades = []

    size = int(input("How many grades (1 - 20): "))

    while size < 1 or size > 20:
        print("Please enter between 1 and 20.")
        size = int(input("How many grades (1 - 20): "))

    for i in range(size):
        grade = int(input(f"Enter grade {i + 1}: "))
        grades.append(grade)

    return grades


def calculate_average(grades):
    total = 0

    for grade in grades:
        total += grade

    return total / len(grades)


def calculate_pass_rate(grades):
    pass_count = 0

    for grade in grades:
        if grade >= 50:
            pass_count += 1

    return (pass_count / len(grades)) * 100


def count_hd(grades):
    hd_count = 0

    for grade in grades:
        if grade >= 80:
            hd_count += 1

    return hd_count


def display_results(average, pass_rate, hd_count):
    print("\nResults")
    print(f"Average Grade: {average:.2f}")
    print(f"Pass Rate: {pass_rate:.2f}%")
    print(f"HD Count: {hd_count}")


def main():
    grades = input_grades()

    average = calculate_average(grades)
    pass_rate = calculate_pass_rate(grades)
    hd_count = count_hd(grades)

    display_results(average, pass_rate, hd_count)

if __name__ == "__main__":
    main()